engine/access/rest/websockets/controller_test.go

```diff
@@ -0,0 +1,282 @@
+	err := controller.configureKeepalive()
+	s.Require().NoError(err, "configureKeepalive should not return an error")
+
+	// Assert expectations
+	s.connection.AssertExpectations(s.T())
+}
+
+// TestControllerShutdown ensures that HandleConnection shuts down gracefully when an error occurs.
+func (s *ControllerSuite) TestControllerShutdown() {
+	s.T().Run("keepalive routine failed", func(*testing.T) {
+		controller := s.initializeController()
+
+		// Mock configureConnection to succeed
+		s.mockConnectionSetup()
+
+		// Mock keepalive to return an error
+		done := make(chan struct{}, 1)
+		s.connection.On("WriteControl", websocket.PingMessage, mock.Anything).Return(func(int, time.Time) error {
+			close(done)
+			return websocket.ErrCloseSent
+		}).Once()
+
+		s.connection.
+			On("ReadJSON", mock.Anything).
+			Return(func(interface{}) error {
+				<-done
+				return websocket.ErrCloseSent
+			}).
+			Once()
+
+		s.connection.On("Close").Return(nil).Once()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		controller.HandleConnection(ctx)
+
+		// Ensure all expectations are met
+		s.connection.AssertExpectations(s.T())
+	})
+
+	s.T().Run("read routine failed", func(*testing.T) {
+		controller := s.initializeController()
+		// Mock configureConnection to succeed
+		s.mockConnectionSetup()
+
+		s.connection.
+			On("ReadJSON", mock.Anything).
+			Return(func(_ interface{}) error {
+				return assert.AnError
+			}).
+			Once()
+
+		s.connection.On("Close").Return(nil).Once()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		controller.HandleConnection(ctx)
+
+		// Ensure all expectations are met
+		s.connection.AssertExpectations(s.T())
+	})
+
+	s.T().Run("write routine failed", func(*testing.T) {
+		controller := s.initializeController()
+
+		// Mock configureConnection to succeed
+		s.mockConnectionSetup()
+		blocksDataProvider := s.mockBlockDataProviderSetup(uuid.New())
+
+		done := make(chan struct{}, 1)
+		requestMessage := models.SubscribeMessageRequest{
+			BaseMessageRequest: models.BaseMessageRequest{Action: "subscribe"},
+			Topic:              dp.BlocksTopic,
+			Arguments:          nil,
+		}
+		msg, err := json.Marshal(requestMessage)
+		s.Require().NoError(err)
+
+		// This is due to how the mock library compares arguments: it requires the
+		// pointers passed in `On` to match the exact memory address of the pointer
+		// passed at runtime. Since these pointers are not guaranteed to be the same,
+		// strict matching (`&msg`) will fail.
+		//
+		// Using `mock.Anything` bypasses this strict matching. The `Run` function
+		// then allows us to simulate the behavior of `ReadJSON` by taking the argument
+		// provided during the method call (a dynamically allocated `*json.RawMessage`)
+		// and setting its value (`*reqMsg = msg`). This ensures that the behavior
+		// mimics the real method while avoiding argument mismatch issues.
+		s.connection.
```

Comment:
I think this is useful context, but the important detail is:
```golang
// Mocks `ReadJSON(v interface{}) error` which accepts an uninitialize interface that
// receives the contents of the read message. This logic mocks that behavior, setting
// the target with the value `msg`
```

Reply:
[cfad79f](https://github.com/onflow/flow-go/pull/6757/commits/cfad79fe97069d6acb12321bbd7666234c263b22)